### PR TITLE
Fixes reflection of hitscan beams against tiles

### PR DESCRIPTION
**Pull request by NotRanged**
## About The Pull Request

When I changed beam reflection code in https://github.com/tgstation/tgstation/pull/56579 I tested it with reflector boxes, but not with wall reflection and reflective blobs.

Turns out, not only does center reflect looks bad on those, it also bugs out and makes the beam hit it. For those tiles, it now properly bounces off the side again and doesn't impact the tile itself.

## Why It's Good For The Game

Blob reflectors were completely broken without this (against anything beam anyway)

## Changelog
:cl:
fix: Fixes hitscan beams not reflecting off of certain tiles (e.g. reflective blob) properly.
/:cl:

### CODE REVIEW

File: code/datums/position_point_vector.dm [hunk 1]
```diff
@@ -151,6 +151,14 @@
 	starting_y = y
 	starting_z = z
 
+/datum/point/vector/proc/set_location(tile_x, tile_y, tile_z, p_x = 0, p_y = 0) // initialize_location of /datum/point/, doesn't set starting_x/y/z
+	if(!isnull(tile_x))
+		x = ((tile_x - 1) * world.icon_size) + world.icon_size / 2 + p_x + 1
+	if(!isnull(tile_y))
+		y = ((tile_y - 1) * world.icon_size) + world.icon_size / 2 + p_y + 1
```
Comment (bobbahbrown) [inline code/datums/position_point_vector.dm:158]: From [contributing guidelines](https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#other-notes): "Do not divide when you can easily convert it to multiplication."

```suggestion
		x = ((tile_x - 1) * world.icon_size) + world.icon_size * 0.5 + p_x + 1
	if(!isnull(tile_y))
		y = ((tile_y - 1) * world.icon_size) + world.icon_size * 0.5 + p_y + 1
```

Comment (NotRanged) [reply code/datums/position_point_vector.dm:158]: I see, I'll scour the file for other needless divisions

File: code/modules/projectiles/projectile.dm [hunk 2]
```diff
@@ -692,7 +692,7 @@
 		START_PROCESSING(SSprojectiles, src)
 	pixel_move(1, FALSE)	//move it now!
 
-/obj/projectile/proc/setAngle(new_angle) //wrapper for overrides.
+/obj/projectile/proc/setAngle(new_angle, center_reflect = FALSE) //wrapper for overrides.
```
Comment (Mothblocks) [inline code/modules/projectiles/projectile.dm:695]: No boolean parameter. Make it a new proc.

File: code/datums/position_point_vector.dm [hunk 3]
```diff
@@ -151,6 +151,14 @@
 	starting_y = y
 	starting_z = z
 
+/datum/point/vector/proc/set_location(tile_x, tile_y, tile_z, p_x = 0, p_y = 0) // initialize_location of /datum/point/, doesn't set starting_x/y/z
```
Comment (Mothblocks) [inline code/datums/position_point_vector.dm:154]:
```suggestion
/// initialize_location of /datum/point/, doesn't set starting_x/y/z
/datum/point/vector/proc/set_location(tile_x, tile_y, tile_z, p_x = 0, p_y = 0)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by NotRanged** [comment]
Made it into a new proc set_angle_centered, renamed SetAngle to set_angle, and turned some comments into codedocs stuff

File: code/datums/position_point_vector.dm [hunk 4]
```diff
@@ -32,7 +33,7 @@
 /datum/position/proc/valid()
 	return x && y && z && !isnull(pixel_x) && !isnull(pixel_y)
 
-/datum/position/New(_x = 0, _y = 0, _z = 0, _pixel_x = 0, _pixel_y = 0) //first argument can also be a /datum/point.
+/datum/position/New(_x = 0, _y = 0, _z = 0, _pixel_x = 0, _pixel_y = 0)	//first argument can also be a /datum/point.
```
Comment (Mothblocks) [inline code/datums/position_point_vector.dm:36]:
```suggestion
/datum/position/New(_x = 0, _y = 0, _z = 0, _pixel_x = 0, _pixel_y = 0) //first argument can also be a /datum/point.
```

File: code/modules/projectiles/projectile.dm [hunk 5]
```diff
@@ -703,15 +703,32 @@
...
-			pcache = trajectory.copy_to()
+		pcache = trajectory.copy_to()
 		store_hitscan_collision(pcache)
 	return TRUE
 
+/// Same as set_angle, but the reflection continues from the center of the object that reflects it instead of the side
+/obj/projectile/proc/set_angle_centered(new_angle)
+	Angle = new_angle
+	if(!nondirectional_sprite)
+		var/matrix/M = new
```
Comment (Mothblocks) [inline code/modules/projectiles/projectile.dm:714]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/modules/projectiles/projectile.dm [hunk 6]
```diff
@@ -703,15 +703,32 @@
...
+		transform = M
+	if(trajectory)
+		trajectory.set_angle(new_angle)
+	
+	var/list/coordinates = trajectory.return_coordinates()
+	trajectory.set_location(coordinates[1], coordinates[2], coordinates[3]) // Sets the trajectory to the center of the tile it bounced at
+
+	if(fired && hitscan && isloc(loc) && (loc != last_angle_set_hitscan_store)) // Handles hitscan projectiles
+		last_angle_set_hitscan_store = loc
+		var/datum/point/pcache = new (src)
```
Comment (Mothblocks) [inline code/modules/projectiles/projectile.dm:725]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Name it `point_cache` or similar.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/position_point_vector.dm (hunk 4 above)
Comment (NotRanged) [reply code/datums/position_point_vector.dm]: In my defense this tab had the exact width of a single spacebar in VSC so I didn't notice it

**Pull request merged**